src/program.cpp

```diff
@@ -805,22 +820,32 @@ cl_build_status cvk_program::compile_source(const cvk_device* device) {
         m_binary_type = CL_PROGRAM_BINARY_TYPE_EXECUTABLE;
     }
 
+    // Translate spirv to llvm ir using llvm-spirv
 #ifdef CLSPV_ONLINE_COMPILER
-    cvk_info("About to compile \"%s\"", build_options.c_str());
-
-    // clspv is based on LLVM. LLVM options parsing is done using global
-    // variable that are not thread safe. Thus, we need to lock call to clspv in
-    // order to ensure a thread safe execution.
-    static std::mutex clspv_compile_lock;
-    clspv_compile_lock.lock();
-    auto result = clspv::CompileFromSourceString(
-        m_source, "", build_options, m_binary.raw_binary(), &m_build_log);
-    clspv_compile_lock.unlock();
+    if (build_from_il) {
+        llvm::LLVMContext Context;
+        llvm::Module* M;
+        std::string Err;
+        auto m_il_start = reinterpret_cast<const unsigned char*>(m_il.data());
+        membuf m_il_buf(m_il_start, m_il_start + m_il.size());
+        std::istream m_il_stream(&m_il_buf);
+        // llvm-spirv is based on LLVM. LLVM options parsing is done using
+        // global variable that are not thread safe. Thus, we need to lock call
+        // to llvm-spirv in order to ensure a thread safe execution.
+        static std::mutex llvmspirv_compile_lock;
+        llvmspirv_compile_lock.lock();
+        if (!llvm::readSpirv(Context, m_il_stream, M, Err)) {
+            cvk_error_fn("Fails to load SPIR-V as LLVM Module: %s", Err);
+            return CL_BUILD_ERROR;
```

Comment:
What happens if we hit this to the std::mutex?  Is it released in the dtor?

Reply:
Ah good catch, I forgot to fix the lock when adding the error check. I'll wrap both uses of mutex with std::lock_guard to prevent future bugs from being introduced as well...